File: app/src/main/java/org/d1scw0rld/bookbag/BookDetailFragment.java
package org.d1scw0rld.bookbag;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.google.android.material.appbar.CollapsingToolbarLayout;

import org.d1scw0rld.bookbag.dto.Book;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

/**
 * A fragment representing a single book detail screen.
 * This fragment is either contained in a {@link MainActivity}
 * in two-pane mode (on tablets) or a {@link BookFragment}
 * on handsets.
 */
public class BookDetailFragment extends Fragment
{
   /**
    * The fragment argument representing the book ID that this fragment
    * represents.
    */
   public static final String BOOK_ID = "book_id";

   private Book book;
   
   private DBAdapter dbAdapter = null;

   private BookDetailFieldsFactory bookDetailFieldsFactory;
   private LinearLayout llCategories;

   @Override
   public void onCreate(Bundle savedInstanceState)
   {
      super.onCreate(savedInstanceState);

      dbAdapter = new DBAdapter(getActivity());
      dbAdapter.open();

      if(getArguments() != null && getArguments().containsKey(BOOK_ID))
      {
         book = dbAdapter.getBook(getArguments().getLong(BOOK_ID));

         bookDetailFieldsFactory = new BookDetailFieldsFactory(getContext(), dbAdapter, book);
      }
   }

   @Override
   public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
   {
      return inflater.inflate(R.layout.fragment_book_detail, container, false);
   }

   @Override
   public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState)
   {
      super.onViewCreated(view, savedInstanceState);

      CollapsingToolbarLayout appBarLayout = requireActivity().findViewById(R.id.toolbar_layout);

      if(appBarLayout != null)
         appBarLayout.setTitle(book.csTitle.value);

      llCategories = view.findViewById(R.id.ll_categories);
   }

   @Override
   public void onPause()
   {
      dbAdapter.close();

      super.onPause();
   }

   @Override
   public void onResume()
   {
      super.onResume();

      dbAdapter.open();

      if (book != null)
         bookDetailFieldsFactory.addFields(llCategories);
   }
}

File: app/src/main/java/org/d1scw0rld/bookbag/BookDetailFragment.kt
package org.d1scw0rld.bookbag

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.appbar.CollapsingToolbarLayout
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.d1scw0rld.bookbag.data.relation.toDto
import org.d1scw0rld.bookbag.databinding.FragmentBookDetailBinding
import org.d1scw0rld.bookbag.ui.state.UiState
import org.d1scw0rld.bookbag.ui.viewmodel.BookDetailViewModel

@AndroidEntryPoint
class BookDetailFragment : BaseFragment() {

    private var _binding: FragmentBookDetailBinding? = null
    private val binding get() = _binding!!

    private val viewModel: BookDetailViewModel by viewModels()

    companion object {
        const val BOOK_ID = "book_id"
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentBookDetailBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val bookIdArg = arguments?.getLong(BOOK_ID) ?: 0L
        viewModel.loadBook(bookIdArg)

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collect { state ->
                    when (state) {
                        is UiState.Loading -> {
                            showProgressBar()
                        }
                        is UiState.Success -> {
                            hideProgressBar()
                            val detailData = state.data
                            val loadedBook = detailData.bookWithFields?.toDto()
                            val ctx = context ?: return@collect

                            // Safe search in parent fragment's view hierarchy
                            val appBarLayout = parentFragment?.view?.findViewById<CollapsingToolbarLayout>(R.id.toolbar_layout)
                            loadedBook?.let { b ->
                                appBarLayout?.title = b.title.value
                            }
                            
                            // Clear categories container first to prevent duplicates on redraw
                            binding.llCategories.removeAllViews()
                            
                            // Construct fields factory locally (safe from memory leaks) using preloaded currencies
                            val fieldsFactory = BookDetailFieldsFactory(ctx, detailData.currencies, loadedBook)
                            fieldsFactory.addFields(binding.llCategories)
                        }
                        is UiState.Error -> {
                            hideProgressBar()
                            Log.e("BookDetailFragment", "Error loading book detail", state.exception)
                        }
                    }
                }
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
